src/VisionModel/resnet.jl

```diff
@@ -0,0 +1,72 @@
+using Flux
+
+basicblock(inplanes::Int, outplanes::Int, downsample::Bool = false) = downsample ? 
+  Chain(Conv((3, 3), inplanes => outplanes, stride = 1, pad = 1),
+        BatchNorm(outplanes, λ = relu),
+        Conv((3, 3), outplanes => outplanes, stride = 2, pad = 1,
+        BatchNorm(outplanes, λ=relu)) : 
+  Chain(Conv((3, 3), inplanes => outplanes, stride = 1, pad = 1),
+        BatchNorm(outplanes, λ = relu),
+        Conv((3, 3), outplanes => outplanes, stride = 1, pad = 1,
+        BatchNorm(outplanes, λ=relu))
+
+bottleneck(inplanes::Int, outplanes::Int, downsample::Bool = false) = downsample ?
+  Chain(Conv((1, 1), inplanes => inplanes, stride = 1),
+        BatchNorm(inplanes, λ = relu),
+        Conv((3, 3), inplanes => inplanes, stride = 1, pad = 1),
+        BatchNorm(inplanes, λ = relu),
+        Conv((1, 1), inplanes => outplanes, stride = 2),
+        BatchNorm(outplanes, λ = relu)) :
+  Chain(Conv((1, 1), inplanes => inplanes, stride = 1),
+        BatchNorm(inplanes, λ = relu),
+        Conv((3, 3), inplanes => inplanes, stride = 1, pad = 1),
+        BatchNorm(inplanes, λ = relu),
+        Conv((1, 1), inplanes => outplanes, stride = 1),
+        BatchNorm(outplanes, λ = relu))
+
+projection(inplanes::Int, outplanes::Int, stride::Int) = Chain(Conv((1, 1), inplanes => outplanes, stride=stride),
+                                                               BatchNorm((outplanes), λ=relu))
+
+identity(inplanes::Int, outplanes::Int, stride::Int) = +
+
+function resnet(channel_config::AbstractArray{Int,1}, block_config::AbstractArray{Int,1})
+  layers = []
+  push!(layers, Conv((7, 7), 3=>inplanes, stride=(2, 2), pad=(3, 3)))
+  push!(layers, BatchNorm(inplanes, λ=relu))
+  push!(layers, MaxPool((3, 3), stride=(2, 2), pad=(1, 1)))
+  inplanes = 64
+  for nrepeats in block_config
+    for i in 1:nrepeats
+      for expansion in channel_config
+        outplanes = inplanes * expansion
+        push!(layers, SkipConnection(block(inplanes, outplanes, i == 1),
+                                     shortcut(inplanes, outplanes, stride)))
+          
+      end
+    end 
+    inplanes *= 2
+  end
+  push!(layers, AdaptiveMeanPool(1, 1))
+  push!(layers, x -> flatten(x, 1))
+  push!(layers, Dense(512 * expansion, 1000))
+
+  Flux.testmode!(layers)
+  return layers
+end
+
+const resnet_config =
+Dict("resnet18" => ([1, 1], [2, 2, 2, 2]),
+     "resnet34" => ([1, 1], [3, 4, 6, 3]),
+     "resnet50" => ([1, 1, 4], [3, 4, 6, 3]),
+     "resnet101" => ([1, 1, 4], [3, 4, 23, 3]),
+     "resnet152" => ([1, 1, 4], [3, 8, 36, 3]))
+
+ResNet18() = resnet(resnet_config["resnet18"]...)
+
+ResNet34() = resnet(resnet_config["resnet34"]...)
+
+ResNet50() = resnet(resnet_config["resnet50"]...)
+
+ResNet101() = resnet(resnet_config["resnet101"]...)
+
+ResNet152() = resnet(resnet_config["resnet152"]...)
```

Comment:
Can you pass in the appropriate values for `block` and `shortcut` arguments? i.e. `basicblock` or `bottleneck` etc.

Reply:
how shall I pass ``shortcut`` blocks?

Reply:
`ResNet50() = resnet(bottleneck, identity, resnet_config["resnet50"]...)`

Reply:
but that will leave the ``projection``

Reply:
Changing our code can resolve this.. the for loops can be while passing the args i.e. during the execution of ``ResNet50()`` and other functions.

Reply:
I don't think any ResNet models use both `identity` and `projection`. It is one or the other. The models that use `bottleneck` use `identity`, and the models that use `basicblock` use `projection`.

Reply:
According to this [link](https://towardsdatascience.com/understanding-and-coding-a-resnet-in-keras-446d7ff84d33) ``ResNet50`` has two different shortcut blocks. Also explained [youtube video](https://www.youtube.com/watch?v=0tBPSxioIZE)

Reply:
It depends on the dimension of the layers as explained [in this article](https://medium.com/@14prakash/understanding-and-implementing-architectures-of-resnet-and-resnext-for-state-of-the-art-image-cf51669e1624). You will find that at last in ``Observations`` header in the link.

Reply:
Yes, appears you are right. So, as described in the [paper](https://www.cv-foundation.org/openaccess/content_cvpr_2016/papers/He_Deep_Residual_Learning_CVPR_2016_paper.pdf) (Section 4.1, "Identity vs. Projection Shortcuts"), we could categorize the shortcuts into three types: A, B, or C.
- A: all shortcuts are identity (w/ zero padding to increase dimensions)
- B: most shortcuts are identity, except projection when increasing dimensions
- C: all shortcuts are projections

So, instead of passing `shortcut` in, we could have `shortcut_config` which can be either `:A`, `:B`, or `:C`. Accordingly, in the loop we can apply the appropriate shortcuts. We'd also have to modify `identity` to be able to pad when necessary (you could do this based on the values of the arguments passed into `identity`).

Reply:
I was looking for algorithms but I didn't find ``all shortcuts are projections`` anywhere. ``all shortcuts are identity`` in vgg-19, right?

Reply:
All resnet models will have ``:B`` config then. I thought of adding : 
```
  for nrepeats in block_config
    for i in 1:nrepeats-1
      for expansion in channel_config
        outplanes = inplanes * expansion
        push!(layers, SkipConnection(block(inplanes, outplanes, i == 1),
                                     identity(inplanes, outplanes, stride)))
      end
    end 
    inplanes *= 2
    push!(layers, SkipConnection(block(inplanes, outplanes, i == 1),
                                   projection(inplanes, outplanes, stride)))    
  end
```

Reply:
The `:A` and `:C` config are in the original paper. VGG doesn't have any shortcuts. For example, ResNet-18 comes with `:A` and `:C` variants. Only the bottleneck ones (50+) exclusively use `:B`.